src/actix/certificate_helpers.rs

```diff
@@ -0,0 +1,138 @@
+    ) -> Result<SslContextHolder, ErrorStack> {
+        let ssl_context = build_ssl_context(tls_config, verify_client_cert)?;
+        Ok(SslContextHolder {
+            ssl_context,
+            tls_config: tls_config.clone(),
+            verify_client_cert,
+            refresh_interval,
+            last_updated: Instant::now(),
+        })
+    }
+
+    fn try_get_ssl_context(&self) -> Option<&SslContext> {
+        if self.last_updated.elapsed() < self.refresh_interval {
+            Some(&self.ssl_context)
+        } else {
+            None
+        }
+    }
+
+    fn refresh_and_get_ssl_context(&mut self) -> Result<&SslContext, ()> {
+        if self.last_updated.elapsed() < self.refresh_interval {
+            Ok(&self.ssl_context)
+        } else {
+            self.refresh()?;
+            Ok(&self.ssl_context)
+        }
+    }
+
+    fn refresh(&mut self) -> Result<(), ()> {
+        log::info!("Refreshing TLS certificates for actix!");
+        match build_ssl_context(&self.tls_config, self.verify_client_cert) {
+            Ok(ssl_context) => {
+                self.ssl_context = ssl_context;
+                Ok(())
+            }
+            Err(_err) => Err(()),
+        }
+    }
+}
+
+pub fn build_ssl_acceptor(settings: &Settings) -> std::io::Result<SslAcceptorBuilder> {
+    let mut acceptor = SslAcceptor::mozilla_modern_v5(SslMethod::tls())?;
+
+    let tls_config = settings
+        .tls
+        .clone()
+        .ok_or_else(Settings::tls_config_is_undefined_error)?;
+
+    let verify_client_cert = settings.service.verify_https_client_certificate;
+    let ssl_context_holder = RwLock::new(SslContextHolder::new(
+        &tls_config,
+        verify_client_cert,
+        Duration::from_secs(tls_config.cert_ttl),
+    )?);
+
+    // Use set_servername_callback to implement dynamic certificate loading and refresh
+    acceptor.set_servername_callback(move |ssl, _alert| -> Result<(), SniError> {
+        let mut should_refresh = true;
+        {
+            let reader = ssl_context_holder.read();
+            if let Some(ssl_context) = reader.try_get_ssl_context() {
+                should_refresh = false;
+                ssl.set_ssl_context(ssl_context).map_err(|error_stack| {
+                    log::error!("Failed to set SSL context: {}", error_stack);
+                    SniError::ALERT_FATAL
+                })?;
+            }
+        }
+
+        if should_refresh {
+            let mut writer = ssl_context_holder.write();
+            let ssl_context = writer
+                .refresh_and_get_ssl_context()
+                .map_err(|_refresh_error| {
+                    log::error!("Failed to refresh certificates!");
+                    SniError::ALERT_FATAL
+                })?;
+            ssl.set_ssl_context(ssl_context).map_err(|error_stack| {
+                log::error!("Failed to set SSL context: {}", error_stack);
+                SniError::ALERT_FATAL
+            })?;
+        }
+        Ok(())
```

Comment:
Though this functions the same, it removes the `should_refresh` boolean:

```suggestion
        {
            let reader = ssl_context_holder.read();
            if let Some(ssl_context) = reader.try_get_ssl_context() {
                ssl.set_ssl_context(ssl_context).map_err(|error_stack| {
                    log::error!("Failed to set SSL context: {}", error_stack);
                    SniError::ALERT_FATAL
                })?;
                return Ok(());
            }
        }

        // If getting the SSL context failed, try to get it again with refreshing
        let mut writer = ssl_context_holder.write();
        let ssl_context = writer
            .refresh_and_get_ssl_context()
            .map_err(|_refresh_error| {
                log::error!("Failed to refresh certificates!");
                SniError::ALERT_FATAL
            })?;
        ssl.set_ssl_context(ssl_context).map_err(|error_stack| {
            log::error!("Failed to set SSL context: {}", error_stack);
            SniError::ALERT_FATAL
        })?;
        Ok(())
```

Reply:
Done! Thanks for the suggestion!